fix(#10056): keep nullable as true for mariadb generated columns

indiVar0508 opened the pull request:
mariaDB does not support setting NOT NULL for generated column
ref: https://mariadb.com/kb/en/generated-columns/#statement-support

added a check in `get_column_specification` for mariadb, checking
if user has not specified nullable set it as True for computed column,
but if user has explicitly set as False raise a compile error.

added testcase for same

<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
<!-- Describe your changes in detail -->

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.

**Have a nice day!**

zzzeek commented:
Hi -

this is a major change so please follow the [pull request template](https://github.com/sqlalchemy/sqlalchemy/blob/main/.github/pull_request_template.md) which here is "short code fix" - we have to approve the approach based on [an issue you'll post]() with a [complete, runnable MCVE](http://stackoverflow.com/help/mcve).     In this case, we have lots of tests for generated columns with mariadb.  why don't they fail?  What test did we miss?   That's the actual bug here.  thanks

zzzeek commented:
OK looks lke you have an issue now, will look

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 201bf319d444dfe7e2316839f3d3dd550ab72323 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 201bf319d444dfe7e2316839f3d3dd550ab72323: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

indiVar0508 commented:
Did a force push as lint job failed due to black reformatting, updated the commit black formatting

CaselIT commented:
Issue actually is https://github.com/sqlalchemy/sqlalchemy/issues/10056

CaselIT commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -2049,6 +2049,19 @@ def visit_set_column_comment(self, create, **kw):
+            #       so mark it as True
+            # TODO: Should raise a warning if user explicitly sets value
+            #       as False?
+            column.nullable = True
```
I think we should change the value only if the user did not specify one itself. 

there is `_user_defined_nullable` on the column that we can look.

indiVar0508 replied:
Updated as per comments, realised `is_mariadb` flag was available so using instead of creating seperate mariadbDDLcompiler class.
.
updated issue number in commit message

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 6ce11430574612170c7a1715c7a64a3180a1269d of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 6ce11430574612170c7a1715c7a64a3180a1269d added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -1063,6 +1063,7 @@ class MyClass(Base):
 from ...sql import sqltypes
 from ...sql import util as sql_util
 from ...sql.compiler import InsertmanyvaluesSentinelOpts
+from ...sql.schema import SchemaConst as SchemaConst
```
**Federico Caselli** (CaselIT) wrote:

no need to use  as SchemaConst

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -1766,7 +1768,21 @@ def visit_regexp_replace_op_binary(self, binary, operator, **kw):
         """Builds column DDL."""
-
+        if self.dialect.is_mariadb is True:
+            if column.computed is not None and column.nullable is not True:
```
**Federico Caselli** (CaselIT) wrote:

we could merge this if with the one above

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

indiVar0508 replied:
Hi @CaselIT , 
Thanks for feedback and review,
Updated this in my local while updating realised,
what should be done if someone explicitly specifies `nullable=None`, should exception be raised or it should be defaulted to True in this case, or like raise exception like saying `please specify nullable as False`?.
currently exception will be raised.

indiVar0508 replied:
Looking into documentation, https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.SchemaConst.NULL_UNSPECIFIED

it seems `nullable=None` should not be treated as nullable so did same and updated the testcase you can review.
Added changelog.

CaselIT replied:
thanks

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

looks mostly ok. Left a couple of comments.

A changelog is also missing

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 29f476d08676e6af7768be18c9a1657eaab1e340 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 29f476d08676e6af7768be18c9a1657eaab1e340 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

CaselIT reviewed (approved):
looks ok to me. I've updated the change on gerrit. Will finish it there.

thanks!

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -1766,7 +1768,21 @@ def visit_regexp_replace_op_binary(self, binary, operator, **kw):
+            if column._user_defined_nullable is SchemaConst.NULL_UNSPECIFIED:
+                column.nullable = True
+            else:
+                raise exc.CompileError(
```
**mike bayer** (zzzeek) wrote:

HMMMMmmmmmm are we anticipating a MariaDB error message?  We typically let the database raise these (this is a very long tradition in SQLAlchemy, there are some exceptions but overall we really try to let the DB tell us what it wants).  MariaDB server is the expert on this error message as well as if it changes someday to support this DDL, so I would remove this.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

sqla-tester commented on test/dialect/mysql/test_compiler.py:
```diff
@@ -558,6 +558,50 @@ def test_column_computed(self, text, persisted):
             "ALWAYS AS (x + 2)%s)" % text,
         )
 
+    @testing.combinations(
```
**mike bayer** (zzzeek) wrote:

I would like to have some round trip tests for this added, basically this exact same test but doing a round trip.  that way we have the assert_raises checking for the database server error as we are removing the CompileError part of this.    This new test can be in mysql/test_query.py and I would make a new suite ComputedTest.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

indiVar0508 replied:
Updated as per comments, moved test to test_query.py and droped explicit compile error

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

I'd like to change the exception raise here and move testing to focus on real server round trips

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -1063,6 +1063,7 @@ class MyClass(Base):
 from ...sql import sqltypes
 from ...sql import util as sql_util
 from ...sql.compiler import InsertmanyvaluesSentinelOpts
+from ...sql.schema import SchemaConst
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -1766,7 +1768,21 @@ def visit_regexp_replace_op_binary(self, binary, operator, **kw):
         """Builds column DDL."""
-
+        if (
+            self.dialect.is_mariadb is True
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

indiVar0508 committed:
fixes: sqlalchemy#10056, handle nullable for generated column in mariadb

mariaDB does not support setting NOT NULL for generated column
ref: https://mariadb.com/kb/en/generated-columns/#statement-support

added a check in `get_column_specification` for mariadb, checking
if user has not specified nullable set it as True for computed column.

added testcase for same
added changelog in unreleased_20
test being done against mysql>=5.7 as generated column is supported
from version 5.7 of mysql

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision c34fbe759295d5320c006dcf547fc2466bbd8681 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset c34fbe759295d5320c006dcf547fc2466bbd8681 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

indiVar0508 commented:
It seems generated column is supported from mysql:5.7 so added `mysql>=5.7` in test suite.

indiVar0508 commented:
Hi, 
Is there any other change expected in this PR, happy to update based on review comments.

indiVar0508 commented:
I see, got it will keep that in mind for future PR(s) :D, 
ig currently this one is failing due to mypy getting upset for a few files for py3.11 unrelated to changes done in this PR, will try to fix them in seperate commit within this PR.
.
also on a tangential query unrelated to this, i wondering do you guys participate in [hactoberfest](https://hacktoberfest.com/participation/) for sqlalchemy (didn't saw any old issue/pr tagged for hactoberfest so just asked.) or will you guys willing to mark this PR under hactorfest event, it will help me in the event.

zzzeek commented:
I can add a tag of some kind here if that's what it needs.   as long as it isnt inviting a ton of PR spam which we've had in the past, possibly in conjunction with this event

zzzeek commented:
it looks like topics are per-project only....

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

indiVar0508 commented:
`it's failing some mysql tests as well. I can re-run the suite to get the current list of what's failing`
I think the new commit was not added to gerrit, [68035e0](https://github.com/sqlalchemy/sqlalchemy/pull/10354/commits/68035e0f77884f361bdcad0514ea731814cb1a34) i have excluded tests for mysql <=5.6 as generated column support was added mysql 5.7.

indiVar0508 commented:
`as long as it isnt inviting a ton of PR spam which we've had in the past, possibly in conjunction with this event` 
Yeah sure, i get it spam could be an issue ig,
np i am also looking into other projects as well, i'll raise some Pr(s) there as well :)

zzzeek commented:
here are the failing tests:

https://jenkins.sqlalchemy.org/job/sqlalchemy_2x_gerrit/cext=cext,db=mysql,pyv=py311/4084/


these will go away after a few days, they are all variants of one test, here's a screenshot

![Screenshot from 2023-10-16 14-45-22](https://github.com/sqlalchemy/sqlalchemy/assets/128223/736c632c-0a30-4367-8b50-b08f8d664194)

indiVar0508 commented:
Yeah @zzzeek ,

I saw those, i think i have fixed them in this commit, [68035e0](https://github.com/sqlalchemy/sqlalchemy/pull/10354/commits/68035e0f77884f361bdcad0514ea731814cb1a34) can you add this in gerrit and try it should pass now.

basically excluding tests for mysql 5.6 as generated column is supported from mysql 5.7

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 68035e0f77884f361bdcad0514ea731814cb1a34 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 68035e0f77884f361bdcad0514ea731814cb1a34 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

zzzeek commented:
> Yeah @zzzeek ,
> 
> I saw those, i think i have fixed them in this commit, [68035e0](https://github.com/sqlalchemy/sqlalchemy/pull/10354/commits/68035e0f77884f361bdcad0514ea731814cb1a34) can you add this in gerrit and try it should pass now.
> 
> basically excluding tests for mysql 5.6 as generated column is supported from mysql 5.7

OK thanks, I had scrolled and didnt see those in there until now, this is a bit of a problem with our workflow...

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -1766,7 +1768,12 @@ def visit_regexp_replace_op_binary(self, binary, operator, **kw):
+            column.nullable = True
         colspec = [
             self.preparer.format_column(column),
             self.dialect.type_compiler_instance.process(
```
**mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

sqla-tester commented on test/dialect/mysql/test_query.py:
```diff
@@ -255,3 +261,47 @@ def test_any_literal(self, connection):
+            Column("y", Integer, Computed("x + 2"), **kwargs),
+        )
+        if connection.engine.dialect.name == "mariadb" and nullable in (
+            False,
```
**mike bayer** (zzzeek) wrote:

hm it looks like "nullable=None" means, on SQL Server at least, "dont write any NULL clause at all".  not sure what it does for mariadb, but it's not really a use case here so this is fine.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

thanks this is great!

federico are you +1 still ?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

* test/dialect/mysql/test_compiler.py (line 561): Done

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

sure

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4869 has been **merged**. Congratulations! :)
